### ISSUE TITLE
How to log `train_loss` with `accelerator.accumulate`?

### ISSUE THREAD

**Issue by takiholadi** [huggingface/accelerate#639]
I want to log `train_loss` at each update step. How to incorporate such logging if using gradient accumulation? Is there any chunk of code for example?

It looks like this is not corect:
```
for step, batch in enumerate(train_dataloader):
    
    with accelerator.accumulate(model):
        outputs = model(**batch)
        loss = outputs.loss
    
        accelerator.backward(loss)
        optimizer.step()
        lr_scheduler.step()
        optimizer.zero_grad()
    
    if accelerator.sync_gradients and accelerator.is_main_process:
        tensorboard_writer.add_scalar('loss_train', loss.item() / args.gradient_accumulation_steps)
```

How to gather the correct loss?

**Comment by muellerzr** [comment]
@takiholadi you'd need to sum all the losses. They already are divided for the number of gradient accumulation steps automatically. E.g.:

```python
avg_loss = 0.
for step, batch in enumerate(train_dataloader):
    
    with accelerator.accumulate(model):
        outputs = model(**batch)
        loss = outputs.loss
        avg_loss += loss.item()
    
        accelerator.backward(loss)
        optimizer.step()
        lr_scheduler.step()
        optimizer.zero_grad()
    
    if accelerator.sync_gradients and accelerator.is_main_process:
        tensorboard_writer.add_scalar('loss_train', avg_loss)
        avg_loss = 0
 
# Or if using the logger:
    if accelerator.sync_gradients:
        accelerator.log({"loss_train":avg_loss})
        avg_loss = 0
```

**Comment by zhang9302002** [comment]
> > [@takiholadi](https://github.com/takiholadi) you'd need to sum all the losses and divide them yourself, since that is done internally during backward.
> > E.g.:
> > avg_loss = 0.
> > for step, batch in enumerate(train_dataloader):
> >     
> >     with accelerator.accumulate(model):
> >         outputs = model(**batch)
> >         loss = outputs.loss
> >         avg_loss += loss.item()
> >     
> >         accelerator.backward(loss)
> >         optimizer.step()
> >         lr_scheduler.step()
> >         optimizer.zero_grad()
> >     
> >     if accelerator.sync_gradients and accelerator.is_main_process:
> >         tensorboard_writer.add_scalar('loss_train', accelerate.gather(avg_loss / accelerator.gradient_accumulation_steps))
> >         avg_loss = 0
> >  
> > # Or if using the logger:
> >     if accelerator.sync_gradients:
> >         accelerator.log({"loss_train":accelerate.gather(avg_loss / accelerator.gradient_accumulation_steps)})
> >         avg_loss = 0
> 
> [@muellerzr](https://github.com/muellerzr) Shouldn't this `accelerate.gather(avg_loss / accelerator.gradient_accumulation_steps)` be rather `accelerator.gather(avg_loss / accelerator.gradient_accumulation_steps).sum()` with `.sum()` at the end ? since `accelerator.gather` returns the values of the tensor across all processes and concatenates them on the first dimension.

Shouldn't it be `accelerator.reduce(avg_loss, reduction="mean") / accelerator.gradient_accumulation_steps` rather than `sum()` ?
